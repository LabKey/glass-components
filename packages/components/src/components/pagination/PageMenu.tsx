import React, { PureComponent, ReactNode } from 'react';
import { DropdownButton, MenuItem } from 'react-bootstrap';

import { blurActiveElement } from '../../util/utils';
import { Tip } from '../..';

interface PageSelectorProps {
    currentPage: number;
    disabled: boolean;
    id: string;
    isFirstPage: boolean;
    isLastPage: boolean;
    loadFirstPage: () => void;
    loadLastPage: () => void;
    pageCount: number;
}

export class PageMenu extends PureComponent<PageSelectorProps> {
    loadFirstPage = (): void => {
        this.props.loadFirstPage();
        blurActiveElement();
    };

    loadLastPage = (): void => {
        this.props.loadLastPage();
        blurActiveElement();
    };

    render(): ReactNode {
        const { currentPage, disabled, id, isFirstPage, isLastPage, pageCount } = this.props;

        return (
            <Tip caption="Current Page" trigger={['hover']}>
                <DropdownButton disabled={disabled} id={`current-page-drop-${id}`} pullRight title={currentPage}>
                    <MenuItem header>Jump To</MenuItem>

                    <MenuItem disabled={disabled || isFirstPage} onClick={this.loadFirstPage}>
                        First Page
                    </MenuItem>

                    <MenuItem disabled={disabled || isLastPage} onClick={this.loadLastPage}>
                        Last Page
                    </MenuItem>

                    <MenuItem header>{disabled ? '...' : pageCount} Total Pages</MenuItem>
                </DropdownButton>
            </Tip>
        );
    }
}
