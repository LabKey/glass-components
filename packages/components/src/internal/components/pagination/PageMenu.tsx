import React, { FC } from 'react';

import { createPortal } from 'react-dom';

import { DropdownButton, MenuDivider, MenuHeader, MenuItem } from '../../dropdowns';
import { useOverlayTriggerState } from '../../OverlayTrigger';
import { Tooltip } from '../../Tooltip';

interface Props {
    currentPage: number;
    disabled: boolean;
    isFirstPage: boolean;
    isLastPage: boolean;
    loadFirstPage: () => void;
    loadLastPage: () => void;
    pageCount: number;
    pageSize: number;
    pageSizes: number[];
    setPageSize: (size: number) => void;
}

export const PageMenu: FC<Props> = props => {
    const {
        currentPage,
        disabled,
        isFirstPage,
        isLastPage,
        loadFirstPage,
        loadLastPage,
        pageCount,
        pageSize,
        pageSizes,
        setPageSize,
    } = props;
    const totalPagesText = disabled ? '...' : `${pageCount} Total Pages`;
    // We have to manually wire up a Tooltip here because we're rendering PageMenu within a btn-group so any extra
    // wrapping elements cause it to render incorrectly.
    const { onMouseEnter, onMouseOut, portalEl, show, targetRef } = useOverlayTriggerState<HTMLDivElement>(
        'page-menu-overlay',
        true,
        false,
        200
    );
    const tooltip = (
        <Tooltip id="view-menu-tooltip" placement="top" targetRef={targetRef}>
            Current Page
        </Tooltip>
    );

    return (
        <DropdownButton
            className="current-page-dropdown"
            disabled={disabled}
            pullRight
            onMouseEnter={onMouseEnter}
            onMouseOut={onMouseOut}
            title={currentPage}
            ref={targetRef}
        >
            <MenuHeader text="Jump To" />
            <MenuItem disabled={disabled || isFirstPage} onClick={loadFirstPage}>
                First Page
            </MenuItem>
            <MenuItem disabled={disabled || isLastPage} onClick={loadLastPage}>
                Last Page
            </MenuItem>
            <MenuHeader className="submenu-footer" text={totalPagesText} />
            <MenuDivider />
            <MenuHeader text="Page Size" />
            {pageSizes?.map(size => (
                <MenuItem key={size} active={size === pageSize} onClick={() => setPageSize(size)}>
                    {size}
                </MenuItem>
            ))}
            {show && createPortal(tooltip, portalEl)}
        </DropdownButton>
    );
};
