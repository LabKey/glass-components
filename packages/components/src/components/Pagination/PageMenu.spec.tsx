import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { PageMenu } from './PageMenu';

describe('PageMenu', () => {
    let props;
    beforeEach(() => {
        props = {
            id: 'model',
            currentPage: 2,
            disabled: false,
            isFirstPage: false,
            isLastPage: false,
            loadFirstPage: jest.fn(),
            loadLastPage: jest.fn(),
            pageCount: 34,
        };
    });

    test('render', () => {
        // Should render with page count 34
        let tree = renderer.create(<PageMenu {...props} />);
        expect(tree.toJSON()).toMatchSnapshot();

        // Should render '...' for pageCount, disabled first/last page MenuItems.
        props.disabled = true;
        tree = renderer.create(<PageMenu {...props} />);
        expect(tree.toJSON()).toMatchSnapshot();

        // First page should be disabled.
        props.disabled = false;
        props.isFirstPage = true;
        expect(tree.toJSON()).toMatchSnapshot();

        // Last page should be disabled
        props.isFirstPage = false;
        props.isLastPage = true;
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('interactions', () => {
        const wrapper = mount<PageMenu>(<PageMenu {...props} />);
        wrapper.find('MenuItem').at(1).find('a').simulate('click');
        expect(props.loadFirstPage).toHaveBeenCalled();
        wrapper.find('MenuItem').at(2).find('a').simulate('click');
        expect(props.loadLastPage).toHaveBeenCalled();
    });
});
