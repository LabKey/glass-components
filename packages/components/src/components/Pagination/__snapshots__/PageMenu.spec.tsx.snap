// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMenu render 1`] = `
<div
  className="dropdown btn-group"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle btn btn-default"
    disabled={false}
    id="current-page-drop-model"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="button"
    type="button"
  >
    2
     
    <span
      className="caret"
    />
  </button>
  <ul
    aria-labelledby="current-page-drop-model"
    className="dropdown-menu dropdown-menu-right"
    role="menu"
  >
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      Jump To
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        First Page
      </a>
    </li>
    <li
      className=""
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Last Page
      </a>
    </li>
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      34
       Total Pages
    </li>
  </ul>
</div>
`;

exports[`PageMenu render 2`] = `
<div
  className="dropdown btn-group"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle btn btn-default"
    disabled={false}
    id="current-page-drop-model"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="button"
    type="button"
  >
    2
     
    <span
      className="caret"
    />
  </button>
  <ul
    aria-labelledby="current-page-drop-model"
    className="dropdown-menu dropdown-menu-right"
    role="menu"
  >
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      Jump To
    </li>
    <li
      className="disabled"
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        First Page
      </a>
    </li>
    <li
      className="disabled"
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Last Page
      </a>
    </li>
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      ...
       Total Pages
    </li>
  </ul>
</div>
`;

exports[`PageMenu render 3`] = `
<div
  className="dropdown btn-group"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle btn btn-default"
    disabled={false}
    id="current-page-drop-model"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="button"
    type="button"
  >
    2
     
    <span
      className="caret"
    />
  </button>
  <ul
    aria-labelledby="current-page-drop-model"
    className="dropdown-menu dropdown-menu-right"
    role="menu"
  >
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      Jump To
    </li>
    <li
      className="disabled"
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        First Page
      </a>
    </li>
    <li
      className="disabled"
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Last Page
      </a>
    </li>
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      ...
       Total Pages
    </li>
  </ul>
</div>
`;

exports[`PageMenu render 4`] = `
<div
  className="dropdown btn-group"
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle btn btn-default"
    disabled={false}
    id="current-page-drop-model"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    role="button"
    type="button"
  >
    2
     
    <span
      className="caret"
    />
  </button>
  <ul
    aria-labelledby="current-page-drop-model"
    className="dropdown-menu dropdown-menu-right"
    role="menu"
  >
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      Jump To
    </li>
    <li
      className="disabled"
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        First Page
      </a>
    </li>
    <li
      className="disabled"
      role="presentation"
    >
      <a
        href="#"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
        tabIndex="-1"
      >
        Last Page
      </a>
    </li>
    <li
      className="dropdown-header"
      onKeyDown={[Function]}
      role="heading"
    >
      ...
       Total Pages
    </li>
  </ul>
</div>
`;
